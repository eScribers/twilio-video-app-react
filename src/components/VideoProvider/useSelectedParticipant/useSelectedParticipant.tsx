import React, { createContext, useContext, useState, useEffect } from 'react';
import { Participant, Room } from 'twilio-video';
import rootStore from '../../../stores';
// import { ROOM_STATE } from '../../../utils/displayStrings';

type selectedParticipantContextType = [Participant | null, (participant: Participant) => void];

export const selectedParticipantContext = createContext<selectedParticipantContextType>(null!);

export default function useSelectedParticipant() {
  const setSelectedParticipant = participant => {
    rootStore.participantsStore.setSelectedParticipant(participant);
  };
  return [rootStore.participantsStore.selectedParticipant, setSelectedParticipant];
}

// type SelectedParticipantProviderProps = {
//   room: Room;
//   children: React.ReactNode;
// };

// export function SelectedParticipantProvider({ room, children }: SelectedParticipantProviderProps) {
//   const [selectedParticipant, _setSelectedParticipant] = useState<Participant | null>(null);
//   const setSelectedParticipant = (participant: Participant) =>
//     _setSelectedParticipant(prevParticipant => (prevParticipant === participant ? null : participant));

//   useEffect(() => {
//     const onDisconnect = () => _setSelectedParticipant(null);
//     room.on(ROOM_STATE.DISCONNECTED, onDisconnect);
//     return () => {
//       room.off(ROOM_STATE.DISCONNECTED, onDisconnect);
//     };
//   }, [room]);

//   return (
//     <selectedParticipantContext.Provider value={[selectedParticipant, setSelectedParticipant]}>
//       {children}
//     </selectedParticipantContext.Provider>
//   );
// }
