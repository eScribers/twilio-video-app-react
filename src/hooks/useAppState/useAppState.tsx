import { useContext } from 'react';
import { StateContext } from '../../state/AppStateProvider';

export function useAppState(): any {
  const context = useContext(StateContext);
  if (!context) {
    throw new Error('useAppState must be used within the AppStateProvider');
  }
  return context;
}
