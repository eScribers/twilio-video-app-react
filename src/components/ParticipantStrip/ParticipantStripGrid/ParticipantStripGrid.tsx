import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Paper from '@material-ui/core/Paper';
import { useEffect, useState } from 'react';
import Grid from '@material-ui/core/Grid';
import useParticipants from '../../../hooks/useParticipants/useParticipants';
import useVideoContext from '../../../hooks/useVideoContext/useVideoContext';
import useSelectedParticipant from '../../VideoProvider/useSelectedParticipant/useSelectedParticipant';
import Participant from '../../Participant/Participant';
const useStyles = makeStyles(theme => ({
  root: {
    flexGrow: 1,
    display: 'flex',
    [theme.breakpoints.down('xs')]: {
      flexWrap: 'wrap',
    },
    flexWrap: 'nowrap',
  },
  paper: {
    padding: theme.spacing(1),
    textAlign: 'center',
    color: theme.palette.text.secondary,
  },
  scrollable: {
    overflow: 'auto',
  },
}));

export interface ParticipantStripGridProps {
  viewMode: string;
}
export default function ParticipantStripGrid({ viewMode }: ParticipantStripGridProps) {
  const {
    room: { localParticipant },
  } = useVideoContext();
  const [currViewMode, setCurrViewMode] = useState('');
  const [lgState, setLgState] = useState<any>(4);
  const [mdState, setMdState] = useState<any>(3);
  const participants = useParticipants();
  const [selectedParticipant, setSelectedParticipant] = useSelectedParticipant();

  useEffect(() => {
    if (currViewMode != viewMode) {
      if (viewMode.includes('3X3')) {
        setMdState(4);
        setLgState(4);
      } else if (viewMode.includes('4X4')) {
        setMdState(3);
        setLgState(3);
      } else if (viewMode.includes('2X2')) {
        setMdState(6);
        setLgState(6);
      }
      setCurrViewMode(viewMode);
    }
  }, [viewMode]);

  const classes = useStyles();

  return (
    <div className={classes.root}>
      <Grid container spacing={3} className={classes.scrollable}>
        <Grid item xs={12} sm={6} md={mdState} lg={lgState}>
          <Paper className={classes.paper}>
            <Participant
              participant={localParticipant}
              isSelected={selectedParticipant === localParticipant}
              onClick={() => setSelectedParticipant(localParticipant)}
            />
          </Paper>
        </Grid>
        {participants.map((participant: any) => (
          <Grid key={participant.sid} item xs={12} sm={6} md={mdState} lg={lgState}>
            <Paper className={classes.paper}>
              <Participant
                key={participant.sid}
                participant={participant}
                isSelected={selectedParticipant === participant}
                onClick={() => setSelectedParticipant(participant)}
              />
            </Paper>
          </Grid>
        ))}
      </Grid>
    </div>
  );
}
