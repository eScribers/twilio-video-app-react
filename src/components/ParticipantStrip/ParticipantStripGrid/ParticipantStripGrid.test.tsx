import EventEmitter from 'events';
import React from 'react';
import ParticipantStripGrid from './ParticipantStripGrid';
import { shallow } from 'enzyme';
import useSelectedParticipant from '../../VideoProvider/useSelectedParticipant/useSelectedParticipant';
import useVideoContext from '../../../hooks/useVideoContext/useVideoContext';
import { useAppState } from '../../../hooks/useAppState/useAppState';

jest.mock('../../../hooks/useVideoContext/useVideoContext');
jest.mock('../../VideoProvider/useSelectedParticipant/useSelectedParticipant');
jest.mock('../../../hooks/useAppState/useAppState');
const mockedVideoContext = useVideoContext as jest.Mock<any>;
const mockUseSelectedParticipant = useSelectedParticipant as jest.Mock<any>;

const mockUseAppState = useAppState as jest.Mock<any>;
mockUseAppState.mockImplementation(() => ({ activeSinkId: '' }));

describe('the ParticipantStrip component', () => {
  mockUseSelectedParticipant.mockImplementation(() => [null, () => {}]);

  it('should correctly render ParticipantInfo components', () => {
    const mockRoom: any = new EventEmitter();
    mockRoom.participants = new Map([
      [0, { sid: 0 }],
      [1, { sid: 1 }],
    ]);
    mockRoom.localParticipant = 'localParticipant';
    mockedVideoContext.mockImplementation(() => ({ room: mockRoom }));
    const wrapper = shallow(<ParticipantStripGrid viewMode={'grid 3x3'} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should add the isSelected prop to the local participant when it is selected', () => {
    mockUseSelectedParticipant.mockImplementation(() => ['localParticipant', () => {}]);
    const mockRoom: any = new EventEmitter();
    mockRoom.participants = new Map([
      [0, { sid: 0 }],
      [1, { sid: 1 }],
    ]);
    mockRoom.localParticipant = 'localParticipant';
    mockedVideoContext.mockImplementation(() => ({ room: mockRoom }));
    const wrapper = shallow(<ParticipantStripGrid viewMode={'grid 3x3'} />);
    expect(
      wrapper
        .find('Participant')
        .at(0)
        .prop('isSelected')
    ).toBe(true);
  });

  it('should add the isSelected prop to the first remote participant when it is selected', () => {
    const mockParticipant = { sid: 0 };
    mockUseSelectedParticipant.mockImplementation(() => [mockParticipant, () => {}]);
    const mockRoom: any = new EventEmitter();
    mockRoom.participants = new Map([
      [0, mockParticipant],
      [1, { sid: 1 }],
    ]);
    mockRoom.localParticipant = 'localParticipant';
    mockedVideoContext.mockImplementation(() => ({ room: mockRoom }));
    const wrapper = shallow(<ParticipantStripGrid viewMode={'grid 3x3'} />);
    expect(
      wrapper
        .find('Participant')
        .at(1)
        .prop('isSelected')
    ).toBe(true);
  });

  it('should correctly render after changing view model', () => {
    const mockParticipant = { sid: 0 };
    mockUseSelectedParticipant.mockImplementation(() => [mockParticipant, () => {}]);
    const mockRoom: any = new EventEmitter();
    mockRoom.participants = new Map([
      [0, mockParticipant],
      [1, { sid: 1 }],
    ]);
    mockRoom.localParticipant = 'localParticipant';
    mockedVideoContext.mockImplementation(() => ({ room: mockRoom }));
    const wrapper = shallow(<ParticipantStripGrid viewMode={'grid 3x3'} />);
    expect(
      wrapper
        .find('Participant')
        .at(1)
        .prop('isSelected')
    ).toBe(true);
  });
});
