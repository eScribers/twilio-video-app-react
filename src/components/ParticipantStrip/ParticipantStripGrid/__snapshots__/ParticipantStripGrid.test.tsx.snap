// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the ParticipantStrip component should correctly render ParticipantInfo components 1`] = `
<div
  className="makeStyles-root-1"
>
  <WithStyles(ForwardRef(Grid))
    className="makeStyles-scrollable-3"
    container={true}
    spacing={3}
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      lg={3}
      md={4}
      sm={6}
      xs={12}
    >
      <WithStyles(ForwardRef(Paper))
        className="makeStyles-paper-2"
      >
        <Participant
          isSelected={false}
          onClick={[Function]}
          participant="localParticipant"
        />
      </WithStyles(ForwardRef(Paper))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      key="0"
      lg={3}
      md={4}
      sm={6}
      xs={12}
    >
      <WithStyles(ForwardRef(Paper))
        className="makeStyles-paper-2"
      >
        <Participant
          isSelected={false}
          key="0"
          onClick={[Function]}
          participant={
            Object {
              "sid": 0,
            }
          }
        />
      </WithStyles(ForwardRef(Paper))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      key="1"
      lg={3}
      md={4}
      sm={6}
      xs={12}
    >
      <WithStyles(ForwardRef(Paper))
        className="makeStyles-paper-2"
      >
        <Participant
          isSelected={false}
          key="1"
          onClick={[Function]}
          participant={
            Object {
              "sid": 1,
            }
          }
        />
      </WithStyles(ForwardRef(Paper))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</div>
`;
